packages/primitives/popover/src/utils/cdk-event.service.ts

```diff
@@ -0,0 +1,166 @@
+        eventTypes.forEach((eventType) => {
+            this.#setPreventPrimitiveFromCdkEvent(primitiveInstance, eventType, true);
+        });
+    }
+
+    allowPrimitiveForCdkMultiEvents<T extends object>(primitiveInstance: T, eventTypes: EventType[]) {
+        eventTypes.forEach((eventType) => {
+            this.#setPreventPrimitiveFromCdkEvent(primitiveInstance, eventType, false);
+        });
+    }
+
+    setPreventPrimitiveFromCdkMixEvents<T extends object>(primitiveInstance: T, eventTypes: PrimitiveConfig) {
+        Object.keys(eventTypes).forEach((eventType) => {
+            this.#setPreventPrimitiveFromCdkEvent(
+                primitiveInstance,
+                eventType as EventType,
+                eventTypes[eventTypeAsPrimitiveConfigKey(eventType as EventType)]
+            );
+        });
+    }
+
+    primitivePreventedFromCdkEvent<T extends object>(primitiveInstance: T, eventType: EventType) {
+        return this.primitiveConfigs?.get(primitiveInstance)?.[eventTypeAsPrimitiveConfigKey(eventType)];
+    }
+
+    addClickDomRootEventCallback(callback: (event: MouseEvent) => void) {
+        this.#clickDomRootEventCallbacks.add(callback);
+    }
+
+    removeClickDomRootEventCallback(callback: (event: MouseEvent) => void) {
+        return this.#clickDomRootEventCallbacks.delete(callback);
+    }
+
+    #setPreventPrimitiveFromCdkEvent<
+        T extends object,
+        R extends EventType,
+        K extends PrimitiveConfig[EventTypeAsPrimitiveConfigKey<R>]
+    >(primitiveInstance: T, eventType: R, value: K) {
+        if (!this.primitiveConfigs?.has(primitiveInstance)) {
+            isDevMode() &&
+                console.error(
+                    '[RdxCdkEventService.preventPrimitiveFromCdkEvent] RDX Primitive instance has not been registered!',
+                    primitiveInstance
+                );
+            return;
+        }
+        switch (eventType) {
+            case 'cdkOverlayOutsideClick':
+                this.primitiveConfigs.get(primitiveInstance)!.preventCdkOverlayOutsideClick = value;
+                break;
+            case 'cdkOverlayEscapeKeyDown':
+                this.primitiveConfigs.get(primitiveInstance)!.preventCdkOverlayEscapeKeyDown = value;
+                break;
+        }
+    }
+
+    #registerOnDestroyCallbacks() {
+        this.destroyRef.onDestroy(() => {
+            this.onDestroyCallbacks.forEach((onDestroyCallback) => onDestroyCallback());
+            this.onDestroyCallbacks.clear();
+        });
+    }
+
+    #listenToClickDomRootEvent() {
+        const target = this.document;
+        const eventName = 'click';
+        const options: boolean | AddEventListenerOptions | undefined = { capture: true };
+        const callback = (event: MouseEvent) => {
+            this.#clickDomRootEventCallbacks.forEach((clickDomRootEventCallback) => clickDomRootEventCallback(event));
+        };
+        target.addEventListener(eventName, callback, options);
```

Comment:
awesome work! 

[in the future](https://github.com/radix-ng/primitives/issues/202) we should use [ `renderer2.listen`](https://angular.dev/api/core/Renderer2#listen) instead of  `target.addEventListener`

we should also add listeners [outside angular zone](https://angular.dev/api/core/NgZone#runOutsideAngular), to avoid calling change-detection

[useful](https://angular.love/running-event-listeners-outside-of-the-ngzone/)

Reply:
Thanks :)

yep, I've seen it there in `@angular/cdk` -> src/cdk/platform/features/backwards-compatibility.ts

```TS
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.dev/license
 */

import {Renderer2, VERSION, ListenerOptions} from '@angular/core';

// TODO(crisbeto): remove this function when making breaking changes for v20.
/**
 * Binds an event listener with specific options in a backwards-compatible way.
 * This function is necessary, because `Renderer2.listen` only supports listener options
 * after 19.1 and during the v19 period we support any 19.x version.
 * @docs-private
 */
export function _bindEventWithOptions(
  renderer: Renderer2,
  target: EventTarget,
  eventName: string,
  callback: (event: any) => boolean | void,
  options: ListenerOptions,
): () => void {
  const major = parseInt(VERSION.major);
  const minor = parseInt(VERSION.minor);

  // Event options in `listen` are only supported in 19.1 and beyond.
  // We also allow 0.0.x, because that indicates a build at HEAD.
  if (major > 19 || (major === 19 && minor > 0) || (major === 0 && minor === 0)) {
    return renderer.listen(target, eventName, callback, options);
  }

  target.addEventListener(eventName, callback, options);

  return () => {
    target.removeEventListener(eventName, callback, options);
  };
}
```

@pimenovoleg Damn! I forgot adding the listeners outside Angular! Good point, thanks :) Will fix it.

Reply:
@pimenovoleg  fixed, does it look better now? :)

Reply:
@pawel-twardziak definitely :)